File: scala/src/main/scala/PrimeFactors.scala
package com.crebma.primeFactors

object PrimeFactors {

  // def factors(num: Int) : List[Int] = {
  //   List[Int]()
  // }


  // def factors(num: Int) : List[Int] = {
  // 	if (num == 1) {
  // 		List[Int]()
  // 	} else {
  // 		List[Int](num)
  // 	}
  // }


  // def factors(num: Int) : List[Int] = {
  // 	if (num == 1) {
  // 		List[Int]()
  // 	} else {
  // 		if (num %2 == 0 && num > 2) {
  // 			List[Int](2, 2)
  // 		} else {
  // 			List[Int](num)
  // 		}
  // 	}
  // }


  // def factors(num: Int) : List[Int] = {
  // 	var primes = List[Int]()
  // 	var number = num
  // 	if (number != 1) {
  // 		while (number % 2 == 0 && number > 2) {
  // 			primes = primes :+ 2
  // 			number /= 2
  // 		}

  // 		while (number % 3 == 0 && number > 3) {
  // 			primes = primes :+ 3
  // 			number /= 3
  // 		}

  // 		if (number != 1) {
  // 			primes = primes :+ number
  // 		}
  // 	}
  // 	primes
  // }


  def factors(num: Int) : List[Int] = {
  	var primes = List[Int]()
  	var number = num
  	for (candidate <- 2 to number) {
  		while (number % candidate == 0) {
  			primes = primes :+ candidate
  			number /= candidate
  		}
  	}
  	primes
  }

}

File: go/src/prime_factors/prime_factors.go
package prime_factors

// func Factors(num int) []int {
//   return []int{}
// }


// func Factors(num int) []int {
//   if num > 1 {
//   	return []int{2}
//   } else {
//   	return []int{}
//   }
// }


func Factors(num int) []int {
	var primes = []int{2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97}
	var results = []int{}
	var remainder = num
	for remainder > 1 {
		for _, prime := range primes {
			if remainder % prime == 0 {
				results = append(results, prime)
				remainder = remainder / prime
			}
		}
	}
	return results
}
